tests/zfs-tests/tests/functional/fault/auto_offline_001_pos.ksh

```diff
@@ -0,0 +1,177 @@
+# source.  A copy of the CDDL is also available via the Internet at
+# http://www.illumos.org/license/CDDL.
+#
+# CDDL HEADER END
+#
+
+#
+# Copyright 2018, loli10K <ezomori.nozomu@gmail.com>. All rights reserved.
+#
+
+. $STF_SUITE/include/libtest.shlib
+. $STF_SUITE/tests/functional/events/events_common.kshlib
+. $STF_SUITE/tests/functional/fault/fault.cfg
+
+#
+# DESCRIPTION:
+# Testing Fault Management Agent ZED Logic - Physically removed device is
+# offlined and onlined when reattached
+#
+# STRATEGY:
+# 1. Create a pool
+# 2. Simulate physical removal of one device
+# 3. Verify the device is offlined
+# 4. Reattach the device
+# 5. Verify the device is onlined
+# 6. Repeat the same tests with a spare device: zed will use the spare to handle
+#    the removed data device
+# 7. Repeat the same tests again with a faulted spare device: zed should offline
+#    the removed data device if no spare is available
+#
+# NOTE: the use of 'block_device_wait' throughout the test helps avoid race
+# conditions caused by mixing creation/removal events from partitioning the
+# disk (zpool create) and events from physically removing it (remove_disk).
+#
+verify_runnable "both"
+
+if is_linux; then
+	# Add one 512b scsi_debug device (4Kn would generate IO errors)
+	# NOTE: must be larger than other "file" vdevs and minimum SPA devsize:
+	# add 32m of fudge
+	load_scsi_debug $(($SPA_MINDEVSIZE/1024/1024+32)) 1 1 1 '512b'
+else
+	log_unsupported "scsi debug module unsupported"
+fi
+
+function cleanup
+{
+	destroy_pool $TESTPOOL
+	rm -f $filedev1
+	rm -f $filedev2
+	rm -f $filedev3
+	rm -f $sparedev
+	unload_scsi_debug
+}
+
+log_assert "ZED detects physically removed devices"
+
+log_onexit cleanup
+
+filedev1="$TEST_BASE_DIR/file-vdev-1"
+filedev2="$TEST_BASE_DIR/file-vdev-2"
+filedev3="$TEST_BASE_DIR/file-vdev-3"
+sparedev="$TEST_BASE_DIR/file-vdev-spare"
+removedev=$(get_debug_device)
+
+typeset poolconfs=("mirror $filedev1 $removedev"
+    "raidz $filedev1 $removedev"
+    "raidz2 $filedev1 $filedev2 $removedev"
+    "raidz3 $filedev1 $filedev2 $filedev3 $removedev"
+    "$filedev1 cache $removedev"
+    "mirror $filedev1 $filedev2 cache $removedev"
+    "raidz $filedev1 $filedev2 $filedev3 cache $removedev"
+)
+
+log_must truncate -s $SPA_MINDEVSIZE $filedev1
+log_must truncate -s $SPA_MINDEVSIZE $filedev2
+log_must truncate -s $SPA_MINDEVSIZE $filedev3
+log_must truncate -s $SPA_MINDEVSIZE $sparedev
+
+for conf in "${poolconfs[@]}"
+do
+	# 1. Create a pool
+	log_must zpool create -f $TESTPOOL $conf
+	block_device_wait
+
+	# 2. Simulate physical removal of one device
+	remove_disk $removedev
+
+	# 3. Verify the device is offlined
+	log_must wait_vdev_state $TESTPOOL $removedev "OFFLINE"
+
+	# 4. Reattach the device
+	insert_disk $removedev
+
+	# 5. Verify the device is onlined
+	log_must wait_vdev_state $TESTPOOL $removedev "ONLINE"
+
+	# cleanup
+	destroy_pool $TESTPOOL
+	log_must parted "/dev/${removedev}" -s -- mklabel msdos
```

Comment:
CentOS6 failure was caused by yet another race condition in udev events processing: the second iteration (first `for` loop) was getting "delete" events for partitions created on the first iteration. I added a `parted / mklabel msdos` and `block_device_wait` at the end of each loop.

Unfortunately we now have the Coverage builder failing "zpool_expand_001_pos": i am going to try and reproduce this on a local Ubuntu17 builder.

Reply:
I wasn't able to reproduce "zpool_expand_001_pos" failure on a Ubuntu17 builder running this changes/PR, but i am seeing a different failure on a Debian9 builder running 6644e5bb6:

```
Test: /usr/share/zfs/zfs-tests/tests/functional/cli_root/zpool_expand/zpool_expand_001_pos (run as root) [00:28] [FAIL]
02:11:46.99 ASSERTION: zpool can be autoexpanded after set autoexpand=on on vdev expansion
02:11:46.99 NOTE: Setting up loopback, scsi_debug, and file vdevs
02:11:46.99 SUCCESS: truncate -s 1073741824 /var/tmp/vdev_lo
02:11:47.02 SUCCESS: losetup /dev/loop3 /var/tmp/vdev_lo
02:11:47.31 SUCCESS: modprobe scsi_debug dev_size_mb=1024 add_host=1 num_tgts=1 max_luns=1 sector_size=512 physblk_exp=0
02:11:48.31 SUCCESS: truncate -s 1073741824 /var/tmp/vdev_raw
02:11:51.02 SUCCESS: zpool create -f -o autoexpand=on testpool1 /dev/loop3 sda /var/tmp/vdev_raw
02:11:51.05 NOTE: Expanding loopback, scsi_debug, and file vdevs
02:11:51.06 SUCCESS: truncate -s 2147483648 /var/tmp/vdev_lo
02:11:51.06 SUCCESS: losetup -c /dev/loop3
02:11:57.66 SUCCESS: truncate -s 2147483648 /var/tmp/vdev_raw
02:12:12.50 SUCCESS: zpool online -e testpool1 /var/tmp/vdev_raw
02:12:12.81 NOTE: testpool1   has previous size: 3019898880 and  expanded size: 6241124352
02:12:13.14 pool testpool1 has not expanded,  1/3 vdevs expanded
02:12:13.14 NOTE: Performing test-fail callback (/usr/share/zfs/zfs-tests/callbacks/zfs_dbgmsg.ksh)
02:12:13.14 =================================================================
02:12:13.14  Tailing last 200 lines of zfs_dbgmsg log
02:12:13.14 =================================================================
[ ... cut previous runs ... ]
02:12:13.15 1541319109   spa_history.c:319:spa_history_log_sync(): txg 4 create pool version 5000; software version zfs-0.8.0-rc1-87-g6644e5bb6; uts linux 4.9.0-3-amd64 #1 SMP Debian 4.9.30-2+deb9u5 (2017-09-19) x86_64
02:12:13.15 1541319109   spa_history.c:319:spa_history_log_sync(): txg 4 set autoexpand=1
02:12:13.15 1541319109   spa_history.c:319:spa_history_log_sync(): txg 4 set feature@async_destroy=enabled
02:12:13.15 1541319109   spa_history.c:319:spa_history_log_sync(): txg 4 set feature@empty_bpobj=enabled
02:12:13.15 1541319109   spa_history.c:319:spa_history_log_sync(): txg 4 set feature@lz4_compress=enabled
02:12:13.15 1541319109   spa_history.c:319:spa_history_log_sync(): txg 4 set feature@multi_vdev_crash_dump=enabled
02:12:13.15 1541319109   spa_history.c:319:spa_history_log_sync(): txg 4 set feature@spacemap_histogram=enabled
02:12:13.15 1541319109   spa_history.c:319:spa_history_log_sync(): txg 4 set feature@enabled_txg=enabled
02:12:13.15 1541319109   spa_history.c:319:spa_history_log_sync(): txg 4 set feature@hole_birth=enabled
02:12:13.15 1541319109   spa_history.c:319:spa_history_log_sync(): txg 4 set feature@extensible_dataset=enabled
02:12:13.15 1541319109   spa_history.c:319:spa_history_log_sync(): txg 4 set feature@embedded_data=enabled
02:12:13.15 1541319109   spa_history.c:319:spa_history_log_sync(): txg 4 set feature@bookmarks=enabled
02:12:13.15 1541319109   spa_history.c:319:spa_history_log_sync(): txg 4 set feature@filesystem_limits=enabled
02:12:13.15 1541319109   spa_history.c:319:spa_history_log_sync(): txg 4 set feature@large_blocks=enabled
02:12:13.15 1541319109   spa_history.c:319:spa_history_log_sync(): txg 4 set feature@large_dnode=enabled
02:12:13.15 1541319109   spa_history.c:319:spa_history_log_sync(): txg 4 set feature@sha512=enabled
02:12:13.15 1541319109   spa_history.c:319:spa_history_log_sync(): txg 4 set feature@skein=enabled
02:12:13.15 1541319109   spa_history.c:319:spa_history_log_sync(): txg 4 set feature@edonr=enabled
02:12:13.15 1541319109   spa_history.c:319:spa_history_log_sync(): txg 4 set feature@userobj_accounting=enabled
02:12:13.15 1541319109   spa_history.c:319:spa_history_log_sync(): txg 4 set feature@encryption=enabled
02:12:13.15 1541319109   spa_history.c:319:spa_history_log_sync(): txg 4 set feature@project_quota=enabled
02:12:13.15 1541319109   spa_history.c:319:spa_history_log_sync(): txg 4 set feature@device_removal=enabled
02:12:13.15 1541319109   spa_history.c:319:spa_history_log_sync(): txg 4 set feature@obsolete_counts=enabled
02:12:13.15 1541319109   spa_history.c:319:spa_history_log_sync(): txg 4 set feature@zpool_checkpoint=enabled
02:12:13.15 1541319109   spa_history.c:319:spa_history_log_sync(): txg 4 set feature@spacemap_v2=enabled
02:12:13.15 1541319109   spa_history.c:319:spa_history_log_sync(): txg 4 set feature@allocation_classes=enabled
02:12:13.15 1541319109   spa_history.c:319:spa_history_log_sync(): txg 4 set feature@resilver_defer=enabled
02:12:13.15 1541319111   spa_history.c:306:spa_history_log_sync(): command: zpool create -f -o autoexpand=on testpool1 /dev/loop3 sda /var/tmp/vdev_raw
02:12:13.15 1541319112   spa_history.c:306:spa_history_log_sync(): command: zfs set mountpoint=/var/tmp/testdir testpool/testfs
02:12:13.15 1541319114   spa.c:7112:spa_async_request(): spa=testpool1 async request task=1
02:12:13.15 1541319129   spa.c:7112:spa_async_request(): spa=testpool1 async request task=1
02:12:13.15 1541319129   spa_history.c:319:spa_history_log_sync(): txg 13 vdev online pool 'testpool1' size: 4093640704(+1073741824)
02:12:13.15 1541319129   spa_history.c:339:spa_history_log_sync(): ioctl reopen
02:12:13.15 1541319129   spa.c:7112:spa_async_request(): spa=testpool1 async request task=1
02:12:13.15 1541319132   spa.c:7112:spa_async_request(): spa=testpool1 async request task=1
02:12:13.15 1541319132   spa_history.c:319:spa_history_log_sync(): txg 18 vdev online pool 'testpool1' size: 6241124352(+2147483648)
02:12:13.15 1541319132   spa_history.c:306:spa_history_log_sync(): command: zpool online -e testpool1 /var/tmp/vdev_raw
02:12:13.15 1541319132   spa_history.c:339:spa_history_log_sync(): ioctl reopen
02:12:13.15 =================================================================
02:12:13.15  End of zfs_dbgmsg log
02:12:13.15 =================================================================
```

`testpool1` (stripe vdev) has successfully expanded from 3019898880 to 6241124352, but the test is only counting 1 expanded device:

> 02:12:13.14 pool testpool1 has not expanded,  **1**/3 vdevs expanded

when it should be counting **2** according to the dbgmsg output (the test greps `zpool history -il`):

> 02:12:13.15 1541319129   spa_history.c:319:spa_history_log_sync(): txg 13 vdev online pool 'testpool1' size: 4093640704(+1073741824)
> 02:12:13.15 1541319132   spa_history.c:319:spa_history_log_sync(): txg 18 vdev online pool 'testpool1' size: 6241124352(+2147483648)

I am going to bump the expansion timeout from 3 to 10 seconds and see what happens to the Coverage builder.

Reply:
That does appear to have helped.  Let's go with the longer time.